docs/data-flow.md

```diff
@@ -0,0 +1,39 @@
+## Data Flow
+
+This shows the general data flow when a user runs facter on the command to lookup facts `a.b` and `c`.
+
+Generally, facter loads fact definitions (`LoadedFact`) to determine all of the things it could collect, including internal (aka core) facts, custom facts (implemented using the `Facter.add` API) and external facts (json, yaml, bash, etc). Each `LoadedFact` specifies a name like `os.family` and a class that can be called later to collect the values, e.g. `Facts::Linux::Os::Release`.
+
+The `QueryParser` parse both user queries `a.b` and `c` and matches each query against all LoadedFacts, returning an array of `SearchedFacts`. These are more like SearchableFacts, since they haven't been searched yet.
+
+Facter attempts to lookup the facts from the cache, otherwise it calls the `InternalFactManager` and `ExternalFactManager` to resolve facts.
+
+For internal facts, facter wraps each `SearchedFact` with a `CoreFact`. The `CoreFact` calls the `call\_the\_resolver` method on the class that the `SearchedFact` references. The `call\_the\_resolver` method then typically delegates to a resolver and returns the fact value which may be scalar or structured data. For example, `os.family` returns a string, but `gce` returns a Hash.
```

Comment:
Do you need to escape the underscores in `call_the_resolver` in this paragraph?

Reply:
ah emacs was getting confused and showing it as italics, I'll fix that